src/codegen/codegen_opengl.cc

```diff
@@ -0,0 +1,270 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file codegen_opengl.cc
+ *
+ * We are targeting OpenGL 3.3. The reason of not targeting a recent version
+ * of OpenGL is to have better compatibility of WebGL 2.
+ */
+#include <tvm/runtime/config.h>
+#include <tvm/packed_func_ext.h>
+#include <vector>
+#include <string>
+#include "./codegen_opengl.h"
+#include "../runtime/thread_storage_scope.h"
+
+namespace tvm {
+namespace codegen {
+
+CodeGenOpenGL::CodeGenOpenGL()
+    : output_(nullptr), inputs_(), iter_var_(nullptr) {
+  // TODO(zhixunt): Implement this.
+  LOG(INFO) << "CodeGenOpenGL::CodeGenOpenGL";
+}
+
+void CodeGenOpenGL::InitFuncState(LoweredFunc f) {
+  CodeGenC::InitFuncState(f);
+  output_ = nullptr;
+  inputs_.clear();
+  iter_var_ = nullptr;
+}
+
+void CodeGenOpenGL::AddFunction(LoweredFunc f) {
+  LOG(INFO) << "CodeGenOpenGL::AddFunction";
+
+  // clear previous generated state.
+  this->InitFuncState(f);
+
+  this->decl_stream << "#version 300 es\n";
+  this->decl_stream << "precision highp float;\n";
+
+  this->decl_stream << "float tvm_get_texel(sampler2D tex, int idx) {\n";
+  this->decl_stream << "  return texelFetch(tex, ivec2(idx, 0), 0).r;\n";
```

Comment:
I'm actually already using GL_R32F. It's in `TextureFormat GetTextureFormat(TVMType type)` in `opengl_device_api.cc`.

Sadly `texelFetch` returns a `vec4` anyway.

Reply:
what is the implication? Does this mean we store data in compact format but the texFetch API will return a duplicated four floats?

Reply:
I think `texelFetch` returns `vec4(the_real_stored_data, 0, 0, 0)`.

Notice that in deeplearnjs, they do something similar:
https://github.com/PAIR-code/deeplearnjs/blob/04fb4353854e6ab9a8f81b25036a819fa675d1ab/src/math/backends/webgl/render_ndarray_gpu_util.ts#L40

They use `texture2D()`, which takes float index instead of int index.

Reply:
gotcha